src/ray/gcs/gcs_server/gcs_resource_report_poller.cc

```diff
@@ -0,0 +1,130 @@
+#include "ray/gcs/gcs_server/gcs_resource_report_poller.h"
+
+namespace ray {
+namespace gcs {
+
+GcsResourceReportPoller::GcsResourceReportPoller(
+    uint64_t max_concurrent_pulls,
+    std::shared_ptr<GcsResourceManager> gcs_resource_manager,
+    std::shared_ptr<rpc::NodeManagerClientPool> raylet_client_pool)
+    : max_concurrent_pulls_(max_concurrent_pulls),
+      inflight_pulls_(0),
+      gcs_resource_manager_(gcs_resource_manager),
+      raylet_client_pool_(raylet_client_pool),
+      poll_period_ms_(boost::posix_time::milliseconds(
+          RayConfig::instance().gcs_resource_report_poll_period_ms())) {}
+
+GcsResourceReportPoller::~GcsResourceReportPoller() { Stop(); }
+
+void GcsResourceReportPoller::Start() {
+  polling_thread_.reset(new std::thread{[this]() {
+    SetThreadName("resource_report_poller");
+    boost::asio::io_service::work work(polling_service_);
+
+    polling_service_.run();
+    RAY_LOG(DEBUG) << "GCSResourceReportPoller has stopped. This should only happen if "
+                      "the cluster has stopped";
+  }});
+}
+
+void GcsResourceReportPoller::Stop() {
+  polling_service_.stop();
+  if (polling_thread_->joinable()) {
+    polling_thread_->join();
+  }
+}
+
+void GcsResourceReportPoller::HandleNodeAdded(
+    std::shared_ptr<rpc::GcsNodeInfo> node_info) {
+  absl::MutexLock guard(&mutex_);
+  const auto node_id = NodeID::FromBinary(node_info->node_id());
+
+  RAY_CHECK(!nodes_.count(node_id)) << "Node with id: " << node_id << " was added twice!";
+
+  auto &state = nodes_[node_id];
+  state.node_id = node_id;
+
+  state.address.set_raylet_id(node_info->node_id());
+  state.address.set_ip_address(node_info->node_manager_address());
+  state.address.set_port(node_info->node_manager_port());
+
+  state.last_pull_time = absl::GetCurrentTimeNanos();
+
+  state.next_pull_timer = std::unique_ptr<boost::asio::deadline_timer>(
+      new boost::asio::deadline_timer(polling_service_));
+
+  polling_service_.post([this, node_id]() { TryPullResourceReport(node_id); });
+}
+
+void GcsResourceReportPoller::HandleNodeRemoved(
+    std::shared_ptr<rpc::GcsNodeInfo> node_info) {
+  NodeID node_id = NodeID::FromBinary(node_info->node_id());
+
+  {
+    absl::MutexLock guard(&mutex_);
+    nodes_.erase(node_id);
+  }
+}
+
+void GcsResourceReportPoller::TryPullResourceReport(const NodeID &node_id) {
+  absl::MutexLock guard(&mutex_);
+
+  to_pull_queue_.push_back(node_id);
```

Comment:
Nit, I think it would be cleaner to push onto the queue outside of this method, and have this method take no args.